docs/src/main/asciidoc/class-loading-reference.adoc

```diff
@@ -0,0 +1,163 @@
+class is used to resolve all the relevant dependencies (both deployment and runtime) that are needed for the Quarkus
+application. The end result of this process is a `CuratedApplication`, which contains all the class loading information
+for the application.
+
+The `CuratedApplication` can then be used to create an `AugmentAction` instance, which can create production application
+and start/restart runtime ones. This application instance exists within an isolated ClassLoader, it is not necessary
+to have any of the Quarkus deployment classes on the class path as the curate process will resolve them for you.
+
+This bootstrap process should be the same no matter how Quarkus is launched, just with different parameters passed in.
+
+=== Current Run Modes
+
+At the moment we have the following use cases for bootstrapping Quarkus:
+
+- Maven creating production application
+- Maven dev mode
+- Gradle creating a production application
+- Gradle dev mode
+- QuarkusTest (Maven, Gradle and IDE)
+- QuarkusUnitTest (Maven, Gradle and IDE)
+- QuarkusDevModeTest (Maven, Gradle and IDE)
+- Arquillian Adaptor
+
+One of the goals of this refactor is to have all these different run modes boot Quarkus in fundamentally the same way.
+
+== ClassLoader Implementations
+
+Quarkus has the following ClassLoaders:
+
+Base ClassLoader:: 
+
+This is usually the normal JVM System ClassLoader. In some environments such as Maven it may be different. This ClassLoader
+is used to load the bootstrap classes, and other ClassLoader instances will delegate the loading of JDK classes to it.
+
+Augment ClassLoader::
+
+This loads all the `-deployment` artifacts and their dependencies, as well as other user dependencies. It does not load the
+application root or any hot deployed code. This ClassLoader is persistent, even if the application restarts it will remain
+(which is why it cannot load application classes that may be hot deployed). It's parent is the base ClassLoader, and it is
+transformer safe.
+
+At present this can be configured to delegate to the Base ClassLoader, however the plan is for this option to go away and
+always have this as an isolated ClassLoader. Making this an isolated ClassLoader is complicated as it means that all
+the builder classes are isolated, which means that use cases that want to customise the build chains are slightly more complex.
+
+Deployment ClassLoader::
+
+This can load all application classes, it's parent is the Augment ClassLoader so it can also load all deployment classes.
+
+This ClassLoader is non-persistent, it will be re-created when the application is started, and is isolated. This ClassLoader
+is the context ClassLoader that is used when running the build steps. It is transformer safe so loading application classes
+will not prevent them from being transformed.
+
+Base Runtime ClassLoader::
+
+This loads all the runtime extension dependencies, as well as other user dependencies. It does not load the
```

Comment:
`as well as other user dependencies` is mentioned here for `Base Runtime Classloader`, however `Augment ClassLoader` also mentions `as well as other user dependencies`.

This seems kind of confusing, but I assume that the former is done at build time and the latter at runtime? Maybe a note should be made about this?